File: Partition Equal Subset Sum.cpp
class Solution {
public:
    bool solve(int pos, int n, vector <int> &nums, int remainingSum, vector <vector <int> > &dp){
        if(pos == n){
            return remainingSum == 0;
        }
        
        if(remainingSum < 0){
            return false;
        }
        
        if(dp[pos][remainingSum] != -1){
            return dp[pos][remainingSum];
        }
        
        return dp[pos][remainingSum] = solve(pos + 1, n, nums, remainingSum, dp) or solve(pos + 1, n, nums, remainingSum - nums[pos], dp);
    }

    bool canPartition(vector<int>& nums) {
        int n = (int) nums.size(), sum = 0;
        for(int i = 0; i < n; ++i){
            sum += nums[i];
        }
        
        vector <vector <int> > dp(n, vector <int> (sum / 2 + 1, -1));
        
        return sum % 2 == 0 and solve(0, n, nums, sum >> 1, dp);
    }
};

File: Partition Equal Subset Sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int totalSum = 0;
        for(int x : nums){
            totalSum += x;
        }
        
        if(totalSum % 2 != 0){
            return false;
        }
        
        boolean[][] canPartition = new boolean[2][1 + totalSum / 2];
        canPartition[0] = new boolean[1 + totalSum / 2];
        
        for(int i = 0; i < nums.length; ++i){
            canPartition[1] = new boolean[1 + totalSum / 2];
            for(int sum = 0; sum <= totalSum / 2; ++sum){
                canPartition[1][sum] = canPartition[0][sum];
                if(sum == nums[i]){
                    canPartition[1][sum] = true;
                }
                else if(sum > nums[i]){
                    canPartition[1][sum] |= canPartition[0][sum - nums[i]];
                }
            }
            canPartition[0] = canPartition[1];
        }
        
        return canPartition[1][totalSum / 2];
    }
}
